### PR TITLE
fix(optimizer): skip subquery optimization if no apply

### PR DESCRIPTION
**Pull request by chenzl25**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- Related https://github.com/risingwavelabs/risingwave/issues/13459

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/frontend/src/optimizer/logical_optimization.rs [hunk 1]
```diff
@@ -426,6 +426,13 @@ impl LogicalOptimizer {
         explain_trace: bool,
         ctx: &OptimizerContextRef,
     ) -> Result<PlanRef> {
+        // In order to unnest a table function, we need to convert it into a `project_set` first.
+        // However, the table function is expected to be converted into a project set, so it must be applied.
```
Comment (xiangjinwu) [inline src/frontend/src/optimizer/logical_optimization.rs:430]: What does this new comment mean?

**Review by xiangjinwu — APPROVED** [review]

File: src/frontend/src/optimizer/logical_optimization.rs [hunk 2]
```diff
@@ -426,6 +426,13 @@ impl LogicalOptimizer {
...
+        plan = plan.optimize_by_rules(&TABLE_FUNCTION_TO_PROJECT_SET);
+        // Bail our if no apply operators.
```
Comment (xxchan) [inline src/frontend/src/optimizer/logical_optimization.rs:432]:
```suggestion
        // Bail out if no apply operators.
```

File: src/frontend/src/optimizer/logical_optimization.rs (hunk 1 above)
Comment (xxchan) [reply src/frontend/src/optimizer/logical_optimization.rs:430]: This rule is also to make table function work in streaming? 🥵

Comment (xxchan) [reply src/frontend/src/optimizer/logical_optimization.rs:430]: BTW, we also discussed whether to abandon the table function operator https://github.com/risingwavelabs/risingwave/pull/11422#pullrequestreview-1564719067

Not sure whether we can/need to put `TABLE_FUNCTION_TO_PROJECT_SET` outside `subquery_unnesting`

Comment (chenzl25) [reply src/frontend/src/optimizer/logical_optimization.rs:430]: I mean this rule it necessary right now. Maybe I should move this rule outside🤔
https://github.com/risingwavelabs/risingwave/blob/4ddcee46cf50b0afc29f979f6816b89d12b63a44/src/frontend/src/optimizer/plan_node/logical_table_function.rs#L124-L140

Comment (chenzl25) [reply src/frontend/src/optimizer/logical_optimization.rs:430]: > This rule is also to make table function work in streaming? 🥵

Yes

**Review by st1page — APPROVED** [review]

**Comment by chenzl25** [comment]
`subquery_unnesting` CPU reduces from 3.9% to 0.1%

**Pull request merged**